Refactored internal fields value cache

czpython opened the pull request:
* Removed all hard coded logic for _{fieldname}_cache.
* Added a common internal API for interacting with the field values cache.

This lays the foundation for a coming fix (in separate pr) to 16043.

Thanks to @carljm and @MarkusH for the support.

charettes commented on django/db/models/query.py:
```diff
@@ -1566,12 +1566,13 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
     # in a dictionary.
 
-    rel_qs, rel_obj_attr, instance_attr, single, cache_name = (
+    rel_qs, rel_obj_attr, instance_attr, single, cache_name, is_cache = (
```
Is there a case where `is_cache is False`?

czpython replied:
Sadly yes https://github.com/django/django/pull/6721/files#diff-50229e6c26e18f27542f698383cdaccfR212.

Without this it's not possible to know when we are passed a field name (that points to descriptor) or the cache name.

charettes replied:
I see thanks, I missed it because of the `is_cached` removal.

carljm replied:
This would be clearer (to me) if the sense were reversed and it were named `is_descriptor` instead of `is_cache`. Would that naming be wrong in some way that I'm missing?

I find this naming confusing because it's named `is_cache` but isn't really describing whether anything is (or is not) "a cache." An alternative to `is_descriptor` if you want to keep the same sense would be something like `is_cache_safe` -- but I still think `is_descriptor` is clearer.

carljm replied:
The other benefit of reversing the sense of this flag is that it makes the "normal" case be `False` and the unusual case `True`, which I think is generally preferable for boolean flags.

czpython replied:
Would you be ok with `is_field`? I like is_descriptor over `is_cache` but seems like it's too implementation specific.

carljm replied:
Hmm. I'm not sure about `is_field` -- don't we generally call them all fields, including FKs etc?

I'm not coming up with anything better than `is_descriptor`. I can see why it feels too "implementation specific", but on the other hand this flag only exists deep inside the implementation of prefetch related. And I think `is_descriptor` accurately captures the conditions under which we need to just set the attribute rather than populating the cache, doesn't it? Can you think of a likely case (even a hypothetical future one) where that term would become too specific?

czpython replied:
Yeah you're right. Better to be explicit about it.
Can't think of a case where it's `True` and not a descriptor

tobiasmcnulty commented on django/db/models/base.py:
```diff
@@ -450,6 +450,14 @@ def __init__(self, db=None):
         # Necessary for correct validation of new instances of objects with explicit (non-auto) PKs.
         # This impacts validation only; it has no effect on the actual save.
         self.adding = True
+        self.cache = {}
```
Previously cached values were internal (prefixed with ``_``). Perhaps this should be too?

tobiasmcnulty replied:
Nevermind, I see this is a separate class.

charettes commented:
This is looking good, I suggest you squash both of your commit into a single one named `Refs #16043 -- Refactored internal fields value cache` and push your second one toward fixing [#16043](https://code.djangoproject.com/ticket/16043) on the same branch. This will make the reviewing process easier IMHO.

carljm commented on django/contrib/contenttypes/fields.py:
```diff
@@ -226,8 +224,8 @@ def __get__(self, instance, cls=None):
         try:
-            rel_obj = getattr(instance, self.cache_attr)
-        except AttributeError:
+            rel_obj = self.get_cached_value(instance)
```
Maybe `get_cached_value` should take an optional second `default` arg (like `dict.get()` does) for cases like this?

czpython replied:
While I agree it makes the API more friendlier, internally None is handled differently than a missing key.
Meaning that None is a valid cache value where as missing key always indicates that the object does not have that field cached.

In the code block above, this would fit in nicely but in most of the other places, there has to be a clear distinction between a missing key and a value of None.

carljm replied:
Adding a default kwarg doesn't mean we have to use it everywhere, and doesn't necessarily mean it defaults to None either. We could default to a sentinel object causing a raise. You would only get a default of None of you explicitly specify one. It's true that would be different from `dict.get`. Still seems potentially worth it to me, but I'm happy to go with your preference.

carljm commented on django/contrib/contenttypes/fields.py:
```diff
@@ -243,7 +241,8 @@ def __get__(self, instance, cls=None):
                 pass
-        setattr(instance, self.cache_attr, rel_obj)
+
+        self.set_cached_value(instance=instance, value=rel_obj)
```
In general, I think the `get` / `set` style APIs are familiar enough that we don't need to use keyword arguments with every call. But I don't feel strongly about it.

czpython replied:
I thought about it, the reason I do it like this is because there are many cases where instance is the value and something else like rel_obj is the instance where the cache is being set. This can look very confusing and inconsistent (in my opinion) if we use positional args.

carljm replied:
Fine by me. I think we could probably omit the keywords in a lot of places and include them only where it's potentially confusing, but probably that's already what you're doing.

carljm commented on django/db/models/base.py:
```diff
@@ -450,6 +450,14 @@ def __init__(self, db=None):
         self.adding = True
+        self.cache = {}
+
+    def set_fields_cache(self, field, value):
```
What is the benefit of having these `set_fields_cache` and `get_fields_cache` methods, over just accessing `instance._state.cache` directly? The only apparent reason is so that you can pass in a field and have `field.get_cache_name()` automatically called, but that could just as well be handled in the field mixin (and it's better handled there, as that keeps the encapsulations cleaner).

carljm replied:
Function calls are also quite expensive in Python, so the burden of proof is to demonstrate that these accessor methods add significant value over just directly accessing the attribute.

czpython replied:
Agree, now that we have the mixin I think it's much better to let that handle the logic with setting field cache.
Regarding `get_fields_cache()`, the reason I thought to leave it in is because as we discussed it's possible to use this cache for other things that are not fields and so I figured it's best to provide a method that gets all fields and returns the dictionary with only field->value vs anything that's not a field. I don't feel too strongly about this though, so I'll go ahead and remove.

carljm replied:
Yeah, I think that falls into YAGNI territory. Better to just do the simpler and faster thing for now, and add indirection later if/when we actually need it.

carljm commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -172,23 +171,26 @@ def __get__(self, instance, cls=None):
         try:
-            rel_obj = getattr(instance, self.cache_name)
-        except AttributeError:
+            rel_obj = self.field.get_cached_value(instance)
```
Another case where a `default` arg would help slightly.

czpython replied:
This is one of the cases I mentioned above, if we replace this with a default of None then we risk running a computation (the block after keyerror) when the cache for rel_obj was explicitly set to None.

carljm replied:
Yeah we don't want that. If we did modify this to use a default value instead of raising, it'd need to be a unique sentinel default, not None. I don't feel strongly that that's an improvement, though.

carljm commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -311,7 +317,7 @@ def RelatedObjectDoesNotExist(self):
 
     def is_cached(self, instance):
-        return hasattr(instance, self.cache_name)
+        return self.related.is_cached(instance)
```
Shouldn't this be `self.related.field.is_cached(instance)`?

And if so, does that reveal a lack of test coverage here that should be fixed?

czpython replied:
This was the source for a lot of confusion for me...

Given the example:

```python
class Restaurant(Model):
            place = OneToOneField(Place, related_name='restaurant')
```

`self.related` is the reverse restaurant field on the `Place` model (`place.restaurant`)
`self.related.field` is the place field on the `Restaurant` model.

So `is_cached()` should check if the reverse `restaurant` field is cached on `Place`

carljm replied:
:+1: We could really use some clearer naming there :)

carljm commented on django/db/models/fields/reverse_related.py:
```diff
@@ -176,9 +177,6 @@ def get_accessor_name(self, model=None):
             return self.related_name
         return opts.model_name + ('_set' if self.multiple else '')
 
-    def get_cache_name(self):
```
How does it work to just remove this method? The default implementation in the mixin returns `self.name`, but `ForeignObjectRel` doesn't appear to have a `name` attribute. Or was this method never actually used?

czpython replied:
`ForeignObjectRel` has a name property but I've noticed it does not do the same as `get_accessor_name()`, will investigate. Good catch!.

carljm commented on django/db/models/fields/utils.py:
```diff
@@ -0,0 +1,26 @@
+    value cache on an instance.
+    """
+
+    def get_cache_name(self):
```
Do we make use of this indirection layer anywhere to set the cache name to something other than `self.name`? It doesn't look like it. If not, maybe we don't actually need the indirection layer at all?

czpython replied:
This was added mostly to keep backwards compatibility and also because ForeignObjectRel used get_accessor_name() but as you pointed out on the comment below, I'm not overriding the get_cache_name(), looks like a bug on my side and also lack of coverage there. Will investigate further.

czpython replied:
So I've gone deeper into it and basically, the reason why this never raised an error is because `self.name` on `ForeignObjectRel` calls `related_query_name` which uses related name or the model name and `get_accessor_name` does the same except that when it's called from a many to one or many to many it adds the `_set` to the default accessor if none is specified using `related_name`.

The cache on the reverse side is only called if the object is a one to one relationship and so in this case `self.name` and `self.get_accessor_name()` will always be the same value.

If it's ok with you I'd like to use `get_accessor_name()` as was the default before, the current tests cover this case already so there's no need to add any as `get_cache_name()` is only called in certain circumstances.

carljm replied:
Yep, seems fine to me.

carljm commented:
This looks a big improvement over the current ad-hoc state of field values caching. Great work! Left some comments suggesting minor improvements.

czpython commented:
@carljm 
Thanks for the review.
I believe I've now addressed all comments and pushed the modifications.

@charettes
I'll wait to see if there's more review adjustments and then push the fix on it's own commit.
Thanks for the suggestion.

carljm commented:
Looks good to me!

timgraham commented:
Did you guys have an offline discussion about the reasoning for putting the cache in model state instead of on the field itself? This adds a sort of weird coupling in my mind where fields now have to know about the implementation of model state. Also, I'm not sure putting the cache there is really in the spirit of that class (e.g. see https://code.djangoproject.com/ticket/19396).

czpython commented:
Yes, the original implementation was to continue to use the model instance with an attribute called ``_fields_cache``. @carljm suggested to use the model state to avoid any potential clashes as the model instance is considered the user space and model state a private space.

I added the `_fields_cache` to model state but later renamed to cache as it was suggested it's better for this to be more general cache, also because model meta already has a `fields_cache`.

timgraham commented:
A more general cache for what? Seems like there may be a risk of cache key collisions depending on what else you put in there?

Instead of putting the methods that deal with model state on `FieldCacheMixin`, maybe they're better suited for `Model`?

Before proceeding with any changes, I'd like to know if @akaariai has any input about the implementation.

carljm commented on django/db/models/query.py:
```diff
@@ -1619,8 +1620,19 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
+                setattr(obj, cache_name, val)
+            else:
+                # No to_attr has been given for this prefetch operation
+                # Using "cache_name" combined with "is_descriptor" means
```
I think this comment needs updating to reflect the reversal of sense from the old `is_field` to the new `is_descriptor`

carljm commented on django/db/models/query.py:
```diff
@@ -1619,8 +1620,19 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
+                setattr(obj, to_attr, val)
+            elif is_descriptor:
+                # cache_name points to a field name in obj
+                # this field is likely a descriptor for a related object.
```
if we are confident enough to name the variable `is_descriptor`, then this comment should be updated to make a stronger assertion?

carljm commented on django/db/models/fields/__init__.py:
```diff
@@ -92,7 +93,7 @@ def _empty(of_cls):
 @total_ordering
 @python_2_unicode_compatible
-class Field(RegisterLookupMixin):
+class Field(RegisterLookupMixin, FieldCacheMixin):
```
Hmm, I just thought of this, but do we really need all Fields to inherit `FieldCacheMixin`? It's really only used by related fields, right? I mean, for normal local fields no cache is needed, the main field attribute on the model is effectively the "cache". Could we get away with having only related fields inherit the mixin?

holvianssi replied:
Yeah, the caching is usually only about related fields.

carljm commented:
Yeah, I suggested using `ModelState`. I don't really see an issue with fields knowing about `ModelState` -- field classes already know lots of other internal details about models (e.g. private APIs like `add_to_class`). (And e.g. ModelForm, which is also an external class, knows about ModelState too). I don't see how https://code.djangoproject.com/ticket/19396 is evidence that this use of ModelState is not "in the spirit" of ModelState -- perhaps you can explain what you mean by that? I see ModelState as a place for ephemeral internal state about a model instance that isn't reflected in its database row; caching field values seems very much in that spirit to me.

All that said, I made the suggestion to use `ModelState` at a point when it looked like we might be adding cache-set and cache-get functions to the `Model` namespace as well, and I was worried about polluting that namespace. Now that it's just a single cache dict, I wouldn't really have a problem with `Model._fields_cache` (it's still using less of that namespace than the existing cache attributes). I don't really see a particularly strong rationale either way, at this point.

I _would_ be concerned about moving the cache-set and cache-get functions onto Model, because that's additional pollution of the Model namespace, which doesn't really "belong" to us. If everything is underscore-prefixed that helps some, but people do add underscore-prefixed methods to their own models. Anytime we add _anything_ to the Model class namespace, it's a potential backwards-incompatibility concern. People who add their own methods to their own Model classes aren't messing with a private API, they're doing something perfectly reasonable and even encouraged.

I also think there's some value in allowing field subclasses to customize how their caching is handled. Having the cache getter and setter on the field class makes that possible.

Regarding naming, I'm also the one who suggested using the name `cache`, though that may have been a mistake. Paulo had originally named it `fields_cache`, but then brought up the potential confusion between `ModelState.fields_cache` and `Options.fields_cache` (which is a cache of field classes, not values). He suggested `ModelState.field_values_cache`, which I thought was unpleasantly verbose, but given that only a few methods on the FieldCacheMixin ever need to use that name, it's fine for it to be verbose. So I think Paulo had the right idea, and I was wrong.

carljm commented:
(Oh, and I would also be very interested in your review of this if you have time, @akaariai !)

czpython commented:
For the record, I'm pro renaming the `cache` attribute to `fields_values_cache`.
I agree with putting the cache in the `ModelState` as suggested by Carl because as far as I can see this class represents a "private" extension of the instance state and as such is qualified to hold instance fields cache, that said I also don't see an issue with having this attribute on the model instance.

Putting the cached values on the fields wouldn't work because field instances are created when the model is loaded (it's possible, just not pretty).

I'll hold off on any changes until we get feedback from @akaariai :)

czpython commented:
@carljm @timgraham Hey guys, is there anything I can do to help speed this one up?

timgraham commented:
Ansii hasn't had time to contribute to Django recently to review this but I think he'll be back before Django 1.11.

carljm commented:
@timgraham I don't think it's fair to contributors to block a PR for over three months waiting for feedback from a specific unavailable core team member. It would be great to get Anssi's thoughts on this if he has time, but we don't do code ownership like that.

IMO this is a nice cleanup that has already gotten review and input from four different core team members (and there's nothing in here that locks us into a corner we can't change later.) If @czpython brings it up to date and addresses all actionable feedback, it should move forward.

holvianssi commented:
I haven't gone into details too much here, but big +1 to the approach in general.

czpython commented:
Hello @carljm @timgraham,
I've rebased with latest master and addressed all comments.

carljm commented on django/db/models/query.py:
```diff
@@ -1581,8 +1582,19 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
+                # No to_attr has been given for this prefetch operation
+                # and the cache_name does not point to a descriptor.
+                # store the value of the field directly in the object's field cache.
+                obj._state.cache[cache_name] = val
```
Poking into the fields cache directly (without going through the usual field-cache API) feels really ugly, as does this whole `is_descriptor` thing. The code here is just much harder to understand than the previous version (unlike in the rest of this PR).

What if we pushed this responsibility back onto the implementation of `get_prefetch_queryset`? Instead of having the last two items of the returned tuple be `cache_name, is_descriptor`, have it be `set_func`, which should be a callable with the signature `(obj, val)`?

Or if that doesn't work out nicely, can't we at least get the field object here and use the normal field-cache-setting API?

czpython replied:
Thanks for the feedback.
I'll have a deeper look to see how this can be improved.

timgraham commented:
@czpython, I noticed some other contributions from you lately so I thought you might have time to finish this. Do you need any guidance?

czpython commented:
Hey @timgraham, thanks for reaching out!
I'm good so far, I got to work on this at PyCon and just have some final touches left. 
I'll have it ready by end of the month.

czpython commented:
@timgraham @carljm @charettes @holvianssi
I've pushed a big update based on Carl's last comment.
Figured that the best way to improve the readibility of this was to refactor the recursive prefetching logic.

The main points are:

* `CacheMixin` - Provides a common API to cache related field values.
* `PrefetchMixin` - Provides a common API for the prefetch mechanism to run prefetches on a field or descriptor.
* `PrefetchQueue` - Holds state for an ongoing prefetch operation. This queue shrinks/expands as prefetch calls are found and executed. This was inspired by @charettes's [comment](https://code.djangoproject.com/ticket/26916#comment:1).

All related descriptors are now prefetcher objects.
Given the models below:

```python
class Author(models.Model):
    name = models.CharField(max_length=200)

class Book(models.Model):
   authors = models.ManyToManyField(Author)
```

Accessing `Book.authors` returns a `ManyToManyDescriptor` descriptor instance, before this pr, the prefetcher object would be accessible via `Book().authors` because the prefetcher was the m2m manager and could only be accessed via an instance. Now the prefetcher is the `ManyToManyDescriptor` descriptor instance, this keeps consistency with the other descriptors (many to one, one to one, etc..) and makes it easier to access the prefecher.

While this PR is fully backwards compatible with all documented behavior and APIs, several components of the prefetching internals have changed, as a result some apps like django-taggit will need to adapt to the new prefetch API. I'll gladly help porting them to be compatible with this change.

Last, I noticed there's an edge-case that's not covered by the test suite but the behavior worked before as a result of [this](https://github.com/django/django/blob/b9f7dce84b7ab5e198129030eae6c1a4aec83d24/django/db/models/query.py#L1515) check.

Calling `prefetch_related_objects` directly with the same `Prefetch` object using a **custom** attribute multiple times will query the database every-time.

```python
books_l = list(Book.objects.all())
# Queries the database
prefetch_related_objects(books_l, Prefetch('authors', to_attr='custom_attr'))
# Queries the database again
prefetch_related_objects(books_l, Prefetch('authors', to_attr='custom_attr'))
```

One of my goals was to remove [these](https://github.com/django/django/blob/b9f7dce84b7ab5e198129030eae6c1a4aec83d24/django/db/models/query.py#L1515) checks but I think in the case of a custom attribute, there aren't many alternatives. I considered adding a new attribute to the model's state that contains a set of custom attributes used on the prefetching.

In that case, calling `prefetch_related_objects(books_l, Prefetch('authors', to_attr='custom_attr'))` would add an entry to ``books_l[0]._state.custom_prefetch_attrs``, this allows us to easily check when a custom attribute has been prefetched, without touching the attribute itself. Also, it would lets us know at an instance level which are custom prefetch attributes, in case these need to be cleared. A drawback with this approach is that it may cause an issue with object size, since the ``custom_prefetch_attrs`` set would be populated for each object in the prefetch.

Let me know what you think, I'm leaning towards adding back the check above and a test for this case.

carljm commented:
@czpython Thanks for all your work here! I took a look through this just now; it looks to me like an improvement in clarity and consistency. I like the unification of prefetchers with the existing descriptors. I definitely did not expect my comment about code clarity in one prefetch method to turn into a deep refactor of all of prefetching, but I like the result!

I know you're going to hate me for this, but this is an awfully big diff now; it changes a lot and it's pretty hard to review with confidence. Do you think it's at all possible to break it down into a couple smaller diffs that can be reviewed and landed independently? It seems like the prior version of the diff, with the fields cache name changed, and ignoring my comment about `is_descriptor`, could be landed as one change, and then the prefetch refactoring as a separate change.

Regarding your question about the edge-case behavior, I'm not sure I fully understand it. I have several questions:
1) Your "this" and "these" links both link to the same place; was there a different link you meant for "these"?
2) If this behavior is currently untested, can we add a test for it?
3) Would checking the instance `__dict__` directly (or using [inspect.getattr_static](https://docs.python.org/3/library/inspect.html#inspect.getattr_static)) help at all here, to allow us to see if some attribute is populated without triggering any kind of code execution?

czpython commented:
Hello @carljm, thanks for your review! Really appreciate it.

I'll force push the cache api changes into this PR (thus removing the new prefetching logic).
Then once this PR lands, I'll open another one with the new prefetching logic and link it to this one for history purposes.

In regards to your questions:

1. No, it was a long message, so I was being redundant.
2. Sure, is it ok to include the test in this pr? or should I add it as part of the big refactor?
3. I wasn't aware of `getattr_static`, I think this might actually solve the problem, I'll double check.

carljm commented:
@czpython re 2) it seems more related to the prefetch refactor, right?

czpython commented:
@carljm You're right, I'll include it there.

czpython commented:
Hello @carljm,
As agreed, I've force pushed to this branch the initial fields cache refactor with the `is_descriptor` logic.

sir-sigurd commented on django/db/models/fields/utils.py:
```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
```
That's not needed anymore.

timgraham replied:
I have this edit and some others ready to push, @czpython can you enable [allow edits form maintainers](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) on this PR?

timgraham replied:
I think I'd name this file `mixins.py`. Seem okay?

czpython replied:
Enabled. Thanks!

timgraham commented on django/db/models/query.py:
```diff
@@ -1707,9 +1722,11 @@ def populate(self, row, from_obj):
-        if obj and self.reverse_cache_name:
-            setattr(obj, self.reverse_cache_name, from_obj)
+
+        self.field.set_cached_value(instance=from_obj, value=obj)
```
Are you passing args as kwargs like this and throughout the patch because of readability? I'm not sure it helps -- it seems natural that a `set()` method would take `(key, value)`.

timgraham commented on django/contrib/contenttypes/fields.py:
```diff
@@ -224,23 +225,22 @@ def __get__(self, instance, cls=None):
             with suppress(ObjectDoesNotExist):
                 rel_obj = ct.get_object_for_this_type(pk=pk_val)
-        setattr(instance, self.cache_attr, rel_obj)
+        self.set_cached_value(instance=instance, value=rel_obj)
```
Are you passing args as kwargs like this and throughout the patch because of readability? I'm not sure it helps -- it seems natural that a `set()` method would take `(key, value)`.

czpython replied:
Yeah, readability was the only reason for this.

czpython commented:
Thanks! @timgraham @charettes @carljm @MarkusH 
What are your thoughts in regards to the second part (prefetch refactor)?
Is it something you'd like to land for 2.0? Just checking to have it ready before the alpha release.

timgraham commented:
Sure, any patches that are ready are candidates for 2.0.
